netplan/libnetplan.py

```diff
@@ -0,0 +1,247 @@
+def netplan_parse(path):
+    # Clear old NetplanNetDefinitions from libnetplan memory
+    lib.netplan_clear_netdefs()
+    err = ctypes.POINTER(_GError)()
+    ret = bool(lib.netplan_parse_yaml(path.encode(), ctypes.byref(err)))
+    if not ret:
+        raise Exception(err.contents.message.decode('utf-8'))
+    lib.netplan_finish_parse(ctypes.byref(err))
+    if err:
+        raise Exception(err.contents.message.decode('utf-8'))
+    return True
+
+
+def _checked_lib_call(fn, *args):
+    err = ctypes.POINTER(_GError)()
+    ret = bool(fn(*args, ctypes.byref(err)))
+    if not ret:
+        raise LibNetplanException(err.contents.message.decode('utf-8'))
+
+
+def netplan_get_filename_by_id(netdef_id, rootdir):
+    res = lib.netplan_get_filename_by_id(netdef_id.encode(), rootdir.encode())
+    return res.decode('utf-8') if res else None
+
+
+class Parser:
+    _abi_loaded = False
+
+    @classmethod
+    def _load_abi(cls):
+        if cls._abi_loaded:
+            return
+
+        lib.netplan_parser_new.restype = _NetplanParserP
+        lib.netplan_parser_clear.argtypes = [ctypes.POINTER(_NetplanParserP)]
+
+        lib.netplan_parser_load_yaml_hierarchy.argtypes = [_NetplanParserP, c_char_p, _GErrorPP]
+        lib.netplan_parser_load_yaml_hierarchy.restype = c_int
+
+        lib.netplan_parser_load_yaml.argtypes = [_NetplanParserP, c_char_p, _GErrorPP]
+        lib.netplan_parser_load_yaml.restype = c_int
+
+        cls._abi_loaded = True
+
+    def __init__(self):
+        self._load_abi()
+        self._ptr = lib.netplan_parser_new()
+
+    def __del__(self):
+        lib.netplan_parser_clear(ctypes.byref(self._ptr))
+
+    def load_yaml_file(self, filename):
+        _checked_lib_call(lib.netplan_parser_load_yaml, self._ptr, filename.encode('utf-8'))
+
+
+class State:
+    _abi_loaded = False
+
+    @classmethod
+    def _load_abi(cls):
+        if cls._abi_loaded:
+            return
+
+        lib.netplan_state_new.restype = _NetplanStateP
+        lib.netplan_state_clear.argtypes = [ctypes.POINTER(_NetplanStateP)]
+
+        lib.netplan_state_import_parser_results.argtypes = [_NetplanStateP, _NetplanParserP, _GErrorPP]
+        lib.netplan_state_import_parser_results.restype = c_int
+
+        lib.netplan_state_get_netdefs_size.argtypes = [_NetplanStateP]
+        lib.netplan_state_get_netdefs_size.restype = c_int
+
+        lib.netplan_state_get_netdef.argtypes = [_NetplanStateP, c_char_p]
+        lib.netplan_state_get_netdef.restype = _NetplanNetDefinitionP
+
+        cls._abi_loaded = True
+
+    def __init__(self):
+        self._load_abi()
+        self._ptr = lib.netplan_state_new()
+
+    def __del__(self):
+        lib.netplan_state_clear(ctypes.byref(self._ptr))
+
+    def import_parser_results(self, parser):
+        _checked_lib_call(lib.netplan_state_import_parser_results, self._ptr, parser._ptr)
+
+    def __len__(self):
+        return lib.netplan_state_get_netdefs_size(self._ptr)
+
+    def __getitem__(self, def_id):
```

Comment:
Nice pythonification, mapping `__len__` and `__getitem__`!

Reply:
Thanks.

An unfortunate side-effect though is that one cannot do `if not state:` to check for state validity anymore, as an empty state will be False-y because its length is 0.